stdlib/public/Concurrency/GlobalExecutor.cpp

```diff
@@ -172,18 +176,77 @@ void swift::donateThreadToGlobalExecutorUntil(bool (*condition)(void *),
 
 #else
 
+// Ensure that Job's layout is compatible with what Dispatch expects.
+// Note: MinimalDispatchObjectHeader just has the fields we care about, it is
+// not complete and should not be used for anything other than these asserts.
+struct MinimalDispatchObjectHeader {
+  const void *VTable;
+  void *Opaque;
+  void *Linkage;
+};
+static_assert(
+    offsetof(Job, metadata) == offsetof(MinimalDispatchObjectHeader, VTable),
+    "Job Metadata field must match location of Dispatch VTable field.");
+static_assert(
+    offsetof(Job, SchedulerPrivate[Job::DispatchLinkage]) ==
+        offsetof(MinimalDispatchObjectHeader, Linkage),
+    "Dispatch Linkage field must be within Job SchedulerPrivate fields.");
+
 /// The function passed to dispatch_async_f to execute a job.
 static void __swift_run_job(void *_job) {
   Job *job = (Job*) _job;
-  swift_job_run(job, ExecutorRef::generic());
+  auto executor =
+      ExecutorRef::fromOpaque(job->SchedulerPrivate[Job::DispatchExecutorRef]);
+  swift_job_run(job, executor);
```

Comment:
Would it make sense to call it something like `forDispatchObject` instead of `forOpaque`, if we do not expect other things to use this opaque constructor? The terms "opaque" and "generic" are quite similar in my mind.

Reply:
That would be good, but this is going away anyway, briefly discussed above with John. I'll use a couple of magic numbers and leave open the possibility of passing through a dispatch queue (not an executor).